pkg/securelaunch/policy/policy.go

```diff
@@ -0,0 +1,227 @@
+ */
+func scanKernelCmdLine() ([]byte, error) {
+
+	slaunch.Debug("scanKernelCmdLine: scanning kernel cmd line for *sl_policy* flag")
+	val, ok := cmdline.Flag("sl_policy")
+	if !ok {
+		log.Printf("scanKernelCmdLine: sl_policy cmdline flag is not set")
+		return nil, errors.New("sl_policy flag not set")
+	}
+
+	// val is of type sda:path
+	mntFilePath, mountPath, e := slaunch.GetMountedFilePath(val, false) // false means readonly mount
+	if e != nil {
+		log.Printf("scanKernelCmdLine: GetMountedFilePath err=%v", e)
+		return nil, fmt.Errorf("scanKernelCmdLine: GetMountedFilePath err=%v", e)
+	}
+	slaunch.Debug("scanKernelCmdLine: Reading file=%s", mntFilePath)
+
+	d, err := ioutil.ReadFile(mntFilePath)
+	if e := mount.Unmount(mountPath, true, false); e != nil {
+		log.Printf("Unmount failed. PANIC")
+		panic(e)
+	}
+
+	if err != nil {
+		// - TODO: should we check for end of file ?
+		log.Printf("Error reading policy file:mountPath=%s, passed=%s", mntFilePath, val)
+		return nil, fmt.Errorf("failed to read policy file:mountPath=%s, passed=%s", mntFilePath, val)
+	}
+	return d, nil
+}
+
+/*
+ *  scanBlockDevice recursively scans an already mounted block device inside directories
+ *	"/", "/efi" and "/boot" for policy file and if found, returns the policy byte as a byte slice.
+ *
+ *	e.g: if you mount /dev/sda1 on /tmp/sda1,
+ *	then mountPath would be /tmp/sda1
+ *	and searchPath would be /tmp/sda1/, /tmp/sda1/efi, and /tmp/sda1/boot
+ *		respectively for each iteration of loop over SearchRoots slice.
+ */
+func scanBlockDevice(mountPath string) ([]byte, bool) {
+
+	log.Printf("scanBlockDevice")
+	// scan for securelaunch.policy under /, /efi, or /boot
+	var SearchRoots = []string{"/", "/efi", "/boot"}
+	for _, c := range SearchRoots {
+
+		searchPath := filepath.Join(mountPath, c)
+		ctx, cancel := context.WithCancel(context.Background())
+		defer cancel()
+		files := find.Find(ctx,
+			find.WithRoot(searchPath),
+			find.WithFilenameMatch("securelaunch.policy"),
+		)
+
+		for o := range files {
+			// Read in policy file:
+			d, err := ioutil.ReadFile(o.Name)
+			if err != nil {
+				log.Printf("Error reading policy file %s, continuing", o.Name)
+				continue
+			}
+			log.Printf("policy file found on mountPath=%s, directory =%s", mountPath, c)
+			return d, true // return when first policy file found
+		}
+		// Policy File not found. Moving on to next search root...
+	}
+	return nil, false
+}
+
+/*
+ * locate() searches for policy file on the kernel cmdline.
+ * if not found on cmdline, it looks for policy file on each block device
+ * under "/", "efi" and "/boot" directories.
+ *
+ * Steps:
+ * 1. Check if kernel param sl_policy is set,
+ * 		parse the string
+ * 2. Iterate through each local block device,
+ *	- mount the block device
+ *	- scan for securelaunch.policy under /, /efi, or /boot
+ * 3  Read in policy file
+ */
+func locate() ([]byte, error) {
+
+	d, _ := scanKernelCmdLine()
```

Comment:
i think you are mixing two different functions...scanKernelCmdLine and scanBlockDevice... The bool returned in scanBlockDevice is checked to see if policy file was found or not which could be checked by just checking if no []byte slice was returned...

Also we dont check the error value of scanKernelCmdLine since this function is an overrride ...i.e. if error is found,
 we do nothing, we simply move on